### PR TITLE
Configurable LDAP user promotion to superuser or staff

### PR DESCRIPTION
**Pull request by joshrivers**
### Pull Request Description

**Overview:**
This pull request refactors settings for LDAP and forward authentication settings in the Django backend. It adds two new files, `fwd_auth_settings.py` and `ldap_settings.py`, to manage authentication-related environment variables and settings based on conditional imports. Additionally, it updates the existing environment check validat to include expectations for the authentication configurations.

**Changes Made:**

1. **New File: ```fwd_auth_settings.py```**
   - Moved forward-auth settings from main settings.py

2. **New File: ```ldap_settings.py```**
   - Moved LDAP settings from main settings.py
   - Defines various LDAP-related configurations, including:
   - Conditional promotion of specific usernames to staff or superuser levels, configured using new environment variables:
      - TA_LDAP_PROMOTE_USERNAMES_TO_SUPERUSER allows a comma-separated list of usernames which will be given staff and superuser permission
      - TA_LDAP_PROMOTE_USERNAMES_TO_STAFF allows a comma-separated list of usernames which will be given only staff permission

3. **Updates to ```ta_envcheck.py```**
   - Added a method ```_additional_vars_expectations()``` to check for the presence of required LDAP environment variables based on the current authentication mode.
   - Adjusts expected and unexpected environment variable checks based on the selected authentication mode.

4. **Modifications to ```settings.py```**
   - Removed the inline LDAP configuration and instead imported settings from ```ldap_settings.py``` when LDAP authentication is enabled.
   - Similarly, imports from ```fwd_auth_settings.py``` when forward authentication is enabled.
   - Fix for edge case where `TA_LOGIN_AUTH_MODE` was set to use LDAP, but `TA_LDAP` was not set.

### CODE REVIEW

**Comment by joshrivers** [comment]
I have been using this compose file as a test harness for standalone validation of the auth settings:
```
name: tubearchivist-testharness
services:
  tubearchivist:
    container_name: tubearchivist
    restart: unless-stopped
    build: .
    ports:
      - 8000:8000
    volumes:
      - media:/youtube
      - cache:/cache
    environment:
      - ES_URL=http://archivist-es:9200 # needs protocol e.g. http and port
      - REDIS_CON=redis://archivist-redis:6379
      - TA_HOST=http://localhost:8000 # set your host name
      - TA_USERNAME=admin
      - TA_PASSWORD=localadminpassword
      - ELASTIC_PASSWORD=vMgTMmpSpCv3kw0LaRbqTXi2kJv4LWa1eCeV9NNZnpNioHS76A5U9qh07 # set password for Elasticsearch
      - TZ=America/New_York
      # - DJANGO_LOG_LEVEL=DEBUG
      # LDAP Settings Block
      - TA_LOGIN_AUTH_MODE=ldap_local
      - TA_LDAP=true
      - TA_LDAP_SERVER_URI=ldap://openldap:1389
      - TA_LDAP_BIND_DN=CN=binduser,ou=users,dc=example,dc=com
      - TA_LDAP_BIND_PASSWORD=bindme
      - TA_LDAP_USER_ATTR_MAP_USERNAME=uid
      - TA_LDAP_USER_BASE=ou=users,dc=example,dc=com
      - TA_LDAP_USER_FILTER=(objectClass=inetOrgPerson)
      - TA_LDAP_PROMOTE_USERNAMES_TO_STAFF=staff01,staff02
      - TA_LDAP_PROMOTE_USERNAMES_TO_SUPERUSER=super01,super02
      - TA_LDAP_USER_ATTR_MAP_PERSONALNAME=displayName
    healthcheck:
      test: [ "CMD", "curl", "-f", "http://localhost:8000/health" ]
      interval: 2m
      timeout: 10s
      retries: 3
      start_period: 30s
    depends_on:
      - archivist-es
      - archivist-redis
    logging:
      options:
        max-size: "100m"
        max-file: "3"
  archivist-redis:
    image: redis
    container_name: archivist-redis
    restart: unless-stopped
    expose:
      - "6379"
    volumes:
      - redis:/data
    depends_on:
      - archivist-es
    logging:
      options:
        max-size: "10m"
        max-file: "3"
  archivist-es:
    image: bbilly1/tubearchivist-es # only for amd64, or use official es 8.13.2
    container_name: archivist-es
    restart: unless-stopped
    environment:
      - "ELASTIC_PASSWORD=vMgTMmpSpCv3kw0LaRbqTXi2kJv4LWa1eCeV9NNZnpNioHS76A5U9qh07" # matching Elasticsearch password
      - "ES_JAVA_OPTS=-Xms1g -Xmx1g"
      - "xpack.security.enabled=true"
      - "discovery.type=single-node"
      - "path.repo=/usr/share/elasticsearch/data/snapshot"
    ulimits:
      memlock:
        soft: -1
        hard: -1
    volumes:
      - es:/usr/share/elasticsearch/data # check for permission error when using bind mount, see readme
    expose:
      - "9200"
    logging:
      options:
        max-size: "50m"
        max-file: "3"
  openldap: # https://hub.docker.com/r/bitnami/openldap
    image: bitnami/openldap:2
    restart: unless-stopped
    configs:
      - source: ldap_bootstrap_ldif
        target: /ldif/bootstrap.ldif
    environment:
      - LDAP_ADMIN_USERNAME=ldapadmin
      - LDAP_ADMIN_PASSWORD=ldapadminpassword
      - LDAP_ROOT=dc=example,dc=com
      - LDAP_CUSTOM_LDIF_DIR=/ldif

configs:
  ldap_bootstrap_ldif:
    content: |
      # Root creation
      dn: dc=example,dc=com
      objectClass: dcObject
      objectClass: organization
      dc: example
      o: example

      dn: ou=users,dc=example,dc=com
      objectClass: organizationalUnit
      ou: users

      # User binduser creation
      dn: cn=binduser,ou=users,dc=example,dc=com
      cn: LDAP Bind User
      sn: Bind
      displayName: Bind User
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: bindme
      uid: binduser
      uidNumber: 1000
      gidNumber: 1000
      homeDirectory: /home/binduser

      # User admin creation
      dn: cn=admin,ou=users,dc=example,dc=com
      cn: Common Admin
      sn: Admin
      displayName: Admin Account
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: adminpassword
      uid: admin
      uidNumber: 1001
      gidNumber: 1001
      homeDirectory: /home/admin

      # User user01 creation
      dn: cn=user01,ou=users,dc=example,dc=com
      cn: User 01
      sn: User
      displayName: Standard User 01
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: password1
      uid: user01
      uidNumber: 1002
      gidNumber: 1002
      homeDirectory: /home/user01

      # User user02 creation
      dn: cn=user02,ou=users,dc=example,dc=com
      cn: User 02
      sn: User
      displayName: Standard User 02
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: password2
      uid: user02
      uidNumber: 1003
      gidNumber: 1003
      homeDirectory: /home/user02

      # User staff01 creation
      dn: cn=staff01,ou=users,dc=example,dc=com
      cn: Staff 01
      sn: Staff
      displayName: Staff User 01
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: password1
      uid: staff01
      uidNumber: 1004
      gidNumber: 1004
      homeDirectory: /home/staff01

      # User staff02 creation
      dn: cn=staff02,ou=users,dc=example,dc=com
      cn: Staff 02
      sn: Staff
      displayName: Staff User 02
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: password2
      uid: staff02
      uidNumber: 1005
      gidNumber: 1005
      homeDirectory: /home/staff02

      # User super01 creation
      dn: cn=super01,ou=users,dc=example,dc=com
      cn: Super 01
      sn: Super
      displayName: Super User 01
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: password1
      uid: super01
      uidNumber: 1006
      gidNumber: 1006
      homeDirectory: /home/super01

      # User super02 creation
      dn: cn=super02,ou=users,dc=example,dc=com
      cn: Super 02
      sn: Super
      displayName: Super User 02
      objectClass: inetOrgPerson
      objectClass: posixAccount
      objectClass: shadowAccount
      userPassword: password2
      uid: super02
      uidNumber: 1007
      gidNumber: 1007
      homeDirectory: /home/super02

      # Viewers Group creation
      dn: cn=viewers,ou=users,dc=example,dc=com
      cn: viewers
      objectClass: groupOfNames
      # User group membership
      member: cn=binduser,ou=users,dc=example,dc=com
      member: cn=admin,ou=users,dc=example,dc=com
      member: cn=user01,ou=users,dc=example,dc=com
      member: cn=user02,ou=users,dc=example,dc=com
      member: cn=staff01,ou=users,dc=example,dc=com
      member: cn=staff02,ou=users,dc=example,dc=com
      member: cn=super01,ou=users,dc=example,dc=com
      member: cn=super02,ou=users,dc=example,dc=com

      # Staff Group creation
      dn: cn=staff,ou=users,dc=example,dc=com
      cn: staff
      objectClass: groupOfNames
      # User group membership
      member: cn=staff01,ou=users,dc=example,dc=com
      member: cn=staff02,ou=users,dc=example,dc=com
      member: cn=super01,ou=users,dc=example,dc=com
      member: cn=super02,ou=users,dc=example,dc=com

      # SuperUser Group creation
      dn: cn=superuser,ou=users,dc=example,dc=com
      cn: superuser
      objectClass: groupOfNames
      # User group membership
      member: cn=super01,ou=users,dc=example,dc=com
      member: cn=super02,ou=users,dc=example,dc=com

volumes:
  cache:
  es:
  media:
  redis:
```
If you down the compose stack and delete volumes between test sessions, you can check the behavior with each set of auth modes and combinations of privilege promotions.

**Review by oblivioncth — CHANGES_REQUESTED** [review]
Just some considerations/refactoring suggestions.

LGTM otherwise, though I haven't had a chance to run the changes yet.

File: backend/config/fwd_auth_settings.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from os import environ
+
+TA_AUTH_PROXY_USERNAME_HEADER = (
+    environ.get("TA_AUTH_PROXY_USERNAME_HEADER") or "HTTP_REMOTE_USER"
```
[the remark this answers is not in the document]
Comment (joshrivers) [reply backend/config/fwd_auth_settings.py:4]: Good catch. Added to my docs PR.

File: backend/config/management/commands/ta_envcheck.py [hunk 2]
```diff
@@ -81,6 +81,7 @@ def handle(self, *args, **options):
         """run all commands"""
         self.stdout.write(LOGO)
         self.stdout.write(TOPIC)
+        self._additional_vars_expectations()
```
[the remark this answers is not in the document]
Comment (joshrivers) [reply backend/config/management/commands/ta_envcheck.py]: Added to my PR. Thanks!

File: backend/config/fwd_auth_settings.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+)
+TA_AUTH_PROXY_LOGOUT_URL = environ.get("TA_AUTH_PROXY_LOGOUT_URL")
```
[the remark this answers is not in the document]
Comment (oblivioncth) [reply backend/config/fwd_auth_settings.py:6]: > I'm totally onboard with you. This was more or less my plan. I actually want to ditch my personal LDAP setup and move to fwdauth, but billy was telling me more people were getting stuck on LDAP, so I've been concentrating there. Getting the test harness compose file up was one of my steps toward working on the fwd code (which I want a similar or shared harness for). Once I have that up, I think it will be pretty smooth to create a similar promotion process for fwdauth users (assuming the Django events are plumbed in that provider...)

Yea, shouldn't be too bad. Kind of you to consider implementing that as well.

> I'm really wishing there was an automated smoke tester to run in those compose stacks to validate each configuration set (either with direct API requests or a Playwright automation). The matrix of regression checks is getting bigger.

Of course it's quite important, but writing/maintaining unit tests is quite tedious. I've never use any kind of "live" testing framework that operates directly on what is essentially an end-user deployment, but writing one that can be run as an Actions workflow would be very fitting for project like this. Something that spins up a test compose stack as you said, and confirms the user can login using every method, with that easily being expandable to cover other functionality.

> On another (later) subject, I'm questioning whether to set up LDAP group support. It would be nice to have, but doesn't add a ton of value to the kinds of users that I'm expecting use TA. It may, however, be a support burden. When I tried it, it was broken with my LDAP provider (Authentik) because of incomplete group query implementation. I don't want to add things which create more support problems than value. 

I also use Authentik and have found groups being available in other applications via OAuth claims to be useful for applications where there's usually a little more granularity with user perms than this one, though even just having a group name specified that if present will make the user an admin (as an alternative method to promote the user could be nice); however, if LDAP in Authentik (and potentially other IdPs) has some edge case issues (I haven't needed to use it yet, only forward-auth and OAuth) then that would be annoying.

Not that it's a great excuse, but I can imagine that the Authentik devs didn't focus on LDAP too much since it's quite old at this point.

**Comment by bbilly1** [comment]
my 2 cents, I'm happy to merge whatever change you suggest here. I honestly don't plan to setup alternative authentication for my selfhosted stuff, but I see how that can be useful.

As long as it doesn't interfere with regular password authentication and there is consensus, I'll merge. 

My recommendations, with lessons learned: 
- Maintainability should be the top priority, not being able to cover all possible edge cases.
- Documentation is another top priority, but looks like you are on track there.

Also I know, we don't have a proper logging setup, something I've been meaning to do, but please let's not mix this in here, that should be a separate thing.

**Comment by joshrivers** [comment]
I can pull the logging code and keep it on my dev branches when I need it.

I've been wondering if I should create a testing folder to contain test harnesses like the docker-compose I put in the discussion above. No strong feeling either way here, but I wondered if it would be useful to you to be able to quickly establish whether the auth configs were broken in code, or just in someone else's configuration.

**Comment by bbilly1** [comment]
I'm going to rely on you fine folks with everything relating to forward auth or LDAP. :-)

on the repo here, I prefer not to have duplicates with compose files, less maintenance needed. 

I also have a separate compose file I use for myself when working on things, I think that's quite common, you can tailor it to your setup and preference.
